### ISSUE TITLE
How to fetch audioSessionId from MediaBrowser.

### ISSUE THREAD

**Issue by prime-zs** [androidx/media#244]
How to fetch audioSessionId from MediaBrowser or tell me some other way to get the audio session id.

**Comment by marcbaechinger** [comment]
The only way to get the audio session ID with a `MediaController`/`MediaBrowser` is registering a `Listener` and implement `onAudioSessionIdChanged(int)`. This callback is currently not called for `MediaController` (see [`MediaControllerImpBase`](https://github.com/androidx/media/blob/release/libraries/session/src/main/java/androidx/media3/session/MediaControllerImplBase.java#L2332)). I think we should call that callback because we promise this by the listener contract. I marked this as a bug.

However, I wonder a bit why you need the audio session ID on the controller side. You can register the same listener on the session/service side when you build your session/player in the service. Can you tell us about your use case of receiving the audio session ID on the controller side?

**Comment by icbaker** [comment]
Drive-by note: `Player.Listener.onAudioSessionIdChanged` will only be invoked if `ExoPlayer.setSessionId` is called - it won't be called with the audio session ID that ExoPlayer generates when it's constructed (this is since [ExoPlayer 2.13.0](https://github.com/google/ExoPlayer/blob/release-v2/RELEASENOTES.md#2130-2021-02-04)).

So although fixing `MediaController` to invoke `Player.Listener.onAudioSessionIdChanged` seems generally correct and worthwhile, I'm not sure if it will actually help to access the auto-generated session ID from the controller side?

**Comment by prime-zs** [comment]
> The only way to get the audio session ID with a `MediaController`/`MediaBrowser` is registering a `Listener` and implement `onAudioSessionIdChanged(int)`. This callback is currently not called for `MediaController` (see [`MediaControllerImpBase`](https://github.com/androidx/media/blob/release/libraries/session/src/main/java/androidx/media3/session/MediaControllerImplBase.java#L2332)). I think we should call that callback because we promise this by the listener contract. I marked this as a bug.
> 
> However, I wonder a bit why you need the audio session ID on the controller side. You can register the same listener on the session/service side when you build your session/player in the service. Can you tell us about your use case of receiving the audio session ID on the controller side?

My requirement was to use it to create an equilizer intent.

**Comment by marcbaechinger** [comment]
Yeah. Sorry I'm  a noob, so you'd have to be a bit more verbose to make me understand.  :)

Generally, I would recommend to move this logic to the session, instead of having it at the controller end. Send a custom command from the controller to the service if you need to initiate this from the controller side. The session can create the `equilizer intent`. Without exactly knowing what this is, I'd say this belongs to the session/service side. 

However, I mark as an enhancement. We will update this bug once there is a change regarding this. I don't think this is of high priority though.

**Comment by prime-zs** [comment]
> Yeah. Sorry I'm a noob, so you'd have to be a bit more verbose to make me understand. :)
> 
> Generally, I would recommend to move this logic to the session, instead of having it at the controller end. Send a custom command from the controller to the service if you need to initiate this from the controller side. The session can create the `equilizer intent`. Without exactly knowing what this is, I'd say this belongs to the session/service side.
> 
> However, I mark as an enhancement. We will update this bug once there is a change regarding this. I don't think this is of high priority though.

Thanks for the suggestion. I was also thinking the same. 
But suppose someone decides to built his own equilizer and want the setting to preview in real time, I guess he might need audioSessionId to do so. Am I right or are their any other solution.
Thanks

**Comment by microkatz** [comment]
@iZakirSheikh 

The onAudioSessionIdChanged events have now been plumbed up to the MediaControllers and can be fetched via a getAudioSessionId method. This was pushed to the main branch in this commit, https://github.com/androidx/media/commit/37449d221f1db19291c02d5dd831355e6f8f44c5. Hopefully that works for your use case!

If not then feel free to re-open or create a new ticket.
